File: 백준/Silver/10799. 쇠막대기/쇠막대기.java
import java.io.*;
import java.util.*;

public class Main{
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int count = 0;
        int line=0;
        String input;
        String[] word;
        String top;
        Stack<String> stack = new Stack<String>();

        input = br.readLine();
        word = input.split("");
        for (String c : word) {
            stack.add(c);
        }
        while(!stack.isEmpty()) {
            top = stack.pop();
            if(!stack.isEmpty()) {
                if(top.equals(")") && stack.peek().equals(")")){
                    line++;
                }
                else if(top.equals(")") && stack.peek().equals("(")){
                    count+=line;
                    stack.pop();
                }
                else{
                    count++;
                    line--;
                }
            }
            else{
                count++;
                line--;
            }
        }
        
        bw.write(count+"\n");
        bw.flush();
        bw.close();
    }
}

File: 백준/Silver/10799. 쇠막대기/쇠막대기.cc
#include <iostream>
#include <stack>

using namespace std;

string str;
int ans = 0, num = 0;
stack<char> s;
int main() {
	cin >> str;
	for (int i = 0; i < str.length(); i++) {
		s.push(str[i]);
	}
	while (!s.empty()) {
		char stackTop = s.top();
		s.pop();
		if (stackTop == ')') {
			if (s.top() == ')') num++;
			else {
				ans += num;
				s.pop();
			}
		}
		else { // stackTop == '('
			ans++; num--;
		}
	}
	cout << ans;
}